### ISSUE TITLE
Set intercept to zero

### ISSUE THREAD

**Issue by yakir12** [JuliaStats/MixedModels.jl#214]
Removing the one term **and** setting the intercept to zero in the following MWE produces an out of memory error every *other* time:
```julia
using Random, MixedModels, DataFrames
rng = MersenneTwister(1234321)
n = 100
x = rand(rng, n)
g = rand(rng, 'a':'z', n)
y = randn(rng, n).*x .+ randn(rng, n)
data = DataFrame(y = y, x = x, g = g)
fm1 = fit!(LinearMixedModel(@formula(y ~ 1 + x + (1|g)), data))
fm2 = fit!(LinearMixedModel(@formula(y ~ 1 + (1|g)), data))
fm3 = fit!(LinearMixedModel(@formula(y ~ 0 + x + (1|g)), data))
fm4 = fit!(LinearMixedModel(@formula(y ~ 0 + (1|g)), data))
```
Namely:
```julia
julia> fm4 = fit!(LinearMixedModel(@formula(y ~ 0 + (1|g)), data))
ERROR: OutOfMemoryError()
Stacktrace:
 [1] Type at ./boot.jl:406 [inlined]
 [2] *(::Adjoint{Float64,MixedModels.FeMat{Float64,Array{Float64,2}}}, ::ReMat{Float64,1}) at /home/yakir/.julia/packages/MixedModels/Vdx1s/src/remat.jl:155
 [3] #LinearMixedModel#11(::Dict{Symbol,Any}, ::Array{Any,1}, ::Type{LinearMixedModel}, ::FormulaTerm{Term,Tuple{ConstantTerm{Int64},FunctionTerm{typeof(|),getfield(Main, Symbol("##235#236")),(:g,)}}}, ::NamedTuple{(:y, :x, :g),Tuple{Array{Float64,1},Array{Float64,1},Array{Char,1}}}) at /home/yakir/.julia/packages/MixedModels/Vdx1s/src/linearmixedmodel.jl:73
 [4] (::getfield(Core, Symbol("#kw#Type")))(::NamedTuple{(:contrasts, :wts),Tuple{Dict{Symbol,Any},Array{Any,1}}}, ::Type{LinearMixedModel}, ::FormulaTerm{Term,Tuple{ConstantTerm{Int64},FunctionTerm{typeof(|),getfield(Main, Symbol("##235#236")),(:g,)}}}, ::NamedTuple{(:y, :x, :g),Tuple{Array{Float64,1},Array{Float64,1},Array{Char,1}}}) at ./none:0
 [5] #LinearMixedModel#10(::Dict{Symbol,Any}, ::Array{Any,1}, ::Type{LinearMixedModel}, ::FormulaTerm{Term,Tuple{ConstantTerm{Int64},FunctionTerm{typeof(|),getfield(Main, Symbol("##235#236")),(:g,)}}}, ::DataFrame) at /home/yakir/.julia/packages/MixedModels/Vdx1s/src/linearmixedmodel.jl:39
 [6] LinearMixedModel(::FormulaTerm{Term,Tuple{ConstantTerm{Int64},FunctionTerm{typeof(|),getfield(Main, Symbol("##235#236")),(:g,)}}}, ::DataFrame) at /home/yakir/.julia/packages/MixedModels/Vdx1s/src/linearmixedmodel.jl:39
 [7] top-level scope at REPL[588]:1
``` 
but when it does work I get:
```julia
julia> fm4 = fit!(LinearMixedModel(@formula(y ~ 0 + (1|g)), data))
Linear mixed model fit by maximum likelihood
 y ~ 0 + (1 | g)
   logLik   -2 logLik     AIC        BIC    
 -158.85588  317.71177  321.71177  326.92211

Variance components:
              Column    Variance  Std.Dev. 
 g        (Intercept)  0.0000000 0.0000000
 Residual              1.4038811 1.1848549
 Number of obs: 100; levels of grouping factors: 26

  Fixed-effects parameters:
───────────────────────────────────────
  Estimate  Std.Error  z value  P(>|z|)
───────────────────────────────────────

───────────────────────────────────────
```
I thought that setting the random seed would remove any randomality here, but I was wrong.

**Comment by palday** [comment]
Can you provide more information about your Julia and package versions? Using e.g. `Pkg.status()`. 

On Julia 1.2, MixedModels `master` and StatsModels 0.6.4 and on Linux x86-64, your first three examples work. 

The last one fails due to a `BoundsError: attempt to access 0×0 Array{Float64,2} at index [0, 0]`, which I'm guessing is the result of trying to update a non-existent fixed effects matrix. That is a genuine bug, albeit in a model that I largely view as degenerate.  

On other thing you could try is to see convert your categorical data to categorical: `categorical!(data, [:g])`. I don't think it will help in your case, but `CategoricalArray` and `PooledArray` are both much more efficient storage for a large number of repetitions of a small number of strings.

**Comment by palday** [comment]
Potentially relevant older comment:

> I'm not entirely sure if the evaluation of the objective works properly when the model matrix for the fixed effects has zero columns, like here.  I'll experiment.

_Originally posted by @dmbates in https://github.com/dmbates/MixedModels.jl/issues/120#issuecomment-448361394_

**Comment by palday** [comment]
The OutOfMemory error seems to have been caused by the way the rank was calculated as part of calculating the pivot (see some of the discussion in #242, where the rank was estimated as `140633143403952`, when it's clearly zero). Now we catch the empty FE as a special case, and the unnecessary and unnecessarily huge allocation doesn't happen. See #253.

**Comment by yakir12** [comment]
I tested the `empty-fe` branch on the code generating this issue, and while it's hard to prove a negative, it didn't error once (tried 10,000 times). So once that's merged this issue could be closed.

**Comment by palday** [comment]
Closed by #253
